astore/cobra.go

```diff
@@ -0,0 +1,60 @@
+package astore
+
+import (
+	"bytes"
+	"context"
+	"github.com/enfabrica/enkit/astore/client/astore"
+	astorepb "github.com/enfabrica/enkit/astore/rpc/astore"
+	"github.com/enfabrica/enkit/lib/client"
+	"github.com/enfabrica/enkit/lib/knetwork"
+	"github.com/enfabrica/enkit/lib/progress"
+	"github.com/spf13/viper"
+	"strings"
+)
+
+type FileFetcher interface {
+	FileContent(ctx context.Context, name string) (string, error)
+}
+
+var _ FileFetcher = (*realFileFetcher)(nil)
+
+type realFileFetcher struct {
+	c astorepb.AstoreClient
+}
+
+func (r *realFileFetcher) FileContent(ctx context.Context, path string) (string, error) {
+	f, err := r.c.Retrieve(ctx, &astorepb.RetrieveRequest{Path: path})
+	if err != nil {
+		return "", err
+	}
+	b := new(strings.Builder)
+	h := progress.NewDiscard()
+	if err := astore.Download(context.TODO(), progress.WriterCreator(h, knetwork.NopWriteCloser(b)), f.Url); err != nil {
+		return "", err
+	}
+	return b.String(), err
+}
+
+func ReadConfig(v *viper.Viper, bf *client.BaseFlags, sf *client.ServerFlags, fileName string, obj interface{}) error {
```

Comment:
Please use the populator library and kflags. Don't roll your own.

Why?
- there is no retry strategy here.
- there is no caching (eg, no papering over config push errors)
- if we change how we do auth, we have to come here and change it.
- no flags to change or override defaults...

Follow what the main here is doing:
https://github.com/enfabrica/enkit/blob/master/astore/client/main.go#L15

@minor-fixes

Reply:
Another major problem is that those configs are different based on the username@domain provided by the user.
If you hard code a path to download the file from, ... we're sort of back to square zero.

Reply:
@ccontavalli can you show me where we can pull in **private** yaml file from astore and populate flags.  I dug through kflags and kcobra for a time and found nothing. I sent a DM on Feb 2nd about it also.

 Also going to address these in order:
 
 * retry is trivial, but can add.
 * we **don't** want local caching because these configs contain api keys which can refresh at a moments notice.  
 * I use all the native functionality from client.BaseFlags and ServerFlags, so any changes to auth will propogate. I don't roll my own at all.  
 * Viper unmarshall doesn't override if the property is not present. This is why I pulled it in. If we want smart merging it's trivial to add also. More than happy to do so.

=== Aside ===

kcobra has some serious overlap with viper, and has some seriously deep rooted bugs that make it not play nice. For example:

* If I registered *client.ServerFlags in this command like astore, the kcobra.populator never touches them. This is why I have to pass in the serverFlags to outputs. This is the case even if I register it with the **same** prefix and command. 
* Any command line parsing from cobra is clobbered flat out. cobra.ExactArgs(1) panics (or errors unexpectedly don't remember).  
* This is my opinion, but the api on commands be the same as it is on projects 50x the size of enkit. `
`
NewCommand().Execute()
` 
 expecting the command to be monkey patched by the flags is an anti pattern.